mem0/memory/main_graph.py

```diff
@@ -1,51 +1,29 @@
 from langchain_community.graphs import Neo4jGraph
-from pydantic import BaseModel, Field
 import json
-from openai import OpenAI
-
-from mem0.embeddings.openai import OpenAIEmbedding
-from mem0.llms.openai import OpenAILLM
+from rank_bm25 import BM25Okapi
+from mem0.utils.factory import LlmFactory, EmbedderFactory
 from mem0.graphs.utils import get_update_memory_messages, EXTRACT_ENTITIES_PROMPT
-from mem0.graphs.tools import UPDATE_MEMORY_TOOL_GRAPH, ADD_MEMORY_TOOL_GRAPH, NOOP_TOOL
-
-client = OpenAI()
-
-class GraphData(BaseModel):
-    source: str = Field(..., description="The source node of the relationship")
-    target: str = Field(..., description="The target node of the relationship")
-    relationship: str = Field(..., description="The type of the relationship")
-
-class Entities(BaseModel):
-    source_node: str
-    source_type: str
-    relation: str
-    destination_node: str
-    destination_type: str
+from mem0.graphs.tools import UPDATE_MEMORY_TOOL_GRAPH, ADD_MEMORY_TOOL_GRAPH, NOOP_TOOL, ADD_MESSAGE_TOOL, SEARCH_TOOL
 
-class ADDQuery(BaseModel):
-    entities: list[Entities]
-
-class SEARCHQuery(BaseModel):
-    nodes: list[str]
-    relations: list[str]
-
-def get_embedding(text):
-    response = client.embeddings.create(
-        model="text-embedding-3-small",
-        input=text
-    )
-    return response.data[0].embedding
 
 class MemoryGraph:
     def __init__(self, config):
         self.config = config
         self.graph = Neo4jGraph(self.config.graph_store.config.url, self.config.graph_store.config.username, self.config.graph_store.config.password)
-
-        self.llm = OpenAILLM()
-        self.embedding_model = OpenAIEmbedding()
+        self.embedding_model = EmbedderFactory.create(
+            self.config.embedder.provider, self.config.embedder.config
+        )
+
+        if self.config.llm.provider:
+            llm_provider = self.config.llm.provider
+        if self.config.graph_store.llm:
+            llm_provider = self.config.graph_store.llm.provider
+        else:
+            llm_provider = "openai_structured"
```

Comment:
This can be set as default in config.

Reply:
Yes, but I am keeping some checks.

1. If the main LLM provider is set then use it. 
2. If the graph LLM provider is ser then use it.
3. If nothing is set then use openai_structured.